### PR TITLE
Use C++17's [[fallthrough]]

### PR DESCRIPTION
**Pull request by masterleinad**
Based on #4232.

Currently, `gcc-7` (`-Wimplicit-fallthrough`) produces tons of warnings because of not handling these fallthrough/break issues. Although there are also warnings from other libraries during build, this PR silences the warnings appearing when compiling user applications.

In the end, we probably need to handle the warnings from other libraries (set a proper `Wimplicit-fallthrough=n` level)

### CODE REVIEW

File: tests/mpi/tria_copy_triangulation.cc [hunk 1]
```diff
@@ -109,11 +112,6 @@ void test()
       else
         Assert (false, ExcInternalError());
     };
-
-//   assert_tria_equal("tria_copy_triangulation", tr, new_tr);
-
-  Assert (tr.n_active_cells() == new_tr.n_active_cells(), ExcInternalError());
-  Assert (tr.n_levels() == new_tr.n_levels(), ExcInternalError());
```
Comment (bangerth) [inline tests/mpi/tria_copy_triangulation.cc]: Any reasons to remove these?

File: include/deal.II/base/table_indices.h [hunk 2]
```diff
@@ -140,22 +140,30 @@ TableIndices<N>::TableIndices(const unsigned int index0,
 
   switch (N)
     {
-    case 1: // fallthrough
+    case 1:
       Assert (index1 == numbers::invalid_unsigned_int, ExcMessage("more than N index values provided"));
-    case 2: // fallthrough
+      DEAL_II_FALLTHROUGH
```
Comment (bangerth) [inline include/deal.II/base/table_indices.h]: To make the life of IDEs easier, what would you think of putting the semicolon here instead of the definition of the macro? If the macro is defined as empty, then this just yields an empty statement. If it is defined as something nonempty, it results in the same as right now. In either case, anything that doesn't know that it's a macro will think of it as a regular statement and indent it correctly.

**Comment by drwells** [comment]
Would you mind splitting this into a separate PR to just add C++17 support? What you have looks right and we can probably get that merged quicker.

File: tests/mpi/tria_copy_triangulation.cc (hunk 1 above)
Comment (masterleinad) [inline tests/mpi/tria_copy_triangulation.cc]: It's just moved up. In `cell1 != tr.end(), cell2 != new_tr.end();` the first statement was never used. However, it should be sufficient to only have one of these two if we first check that the number of active cells is the same.
Let me extract this in another PR.

**Comment by masterleinad** [comment]
@drwells Sure, I'll split this up if the C++17 support part already looks good to you.

File: tests/mpi/tria_copy_triangulation.cc (hunk 1 above)
Comment (masterleinad) [inline tests/mpi/tria_copy_triangulation.cc]: Moved to #4231

File: include/deal.II/base/table_indices.h (hunk 2 above)
Comment (masterleinad) [inline include/deal.II/base/table_indices.h]: Fine with me. At least `astyle` seems to already handle this correctly in either form.

**Comment by bangerth** [comment]
You'll have to rebase this one now that #4232 is merged.

**Comment by masterleinad** [comment]
Rebased
